### PR TITLE
[generator] Add enum support for [FieldAttribute], Fixes 57537 and 57535.

### PR DESCRIPTION
**Pull request by dalexsoto**
Fixes Bugs:
* https://bugzilla.xamarin.com/show_bug.cgi?id=57535
* https://bugzilla.xamarin.com/show_bug.cgi?id=57537

New Features:
* Added Smart Enum support to [Field]
* Added [Native] enum support to [Field]
* Added enum support to [Field]
* Added uint support to [Field]

Also fixed a bug in Smart Enums support when computing the library name
taking into account third party bindings and some special cases, this is
needed because we ended up generating invalid code when Smart Enum support
was used in third party bindings.

Also added Field Enum generation test.

Build diff: https://gist.github.com/dalexsoto/2b30972e53e7fe3d51ac168dd22627d3

### CODE REVIEW

File: src/ObjCRuntime/Dlfcn.cs [hunk 1]
```diff
@@ -222,11 +238,35 @@ public static nint GetNInt (IntPtr handle, string symbol)
 			return (nint)GetIntPtr (handle, symbol);
 		}
 
+		public static void SetNInt (IntPtr handle, string symbol, nint value)
+		{
+			var indirect = dlsym (handle, symbol);
```
Comment (rolfbjarne) [inline src/ObjCRuntime/Dlfcn.cs]: You can simplify this by just doing:

```csharp
SetIntPtr (handle, symbol, (IntPtr) value);
```

Same for `SetNUInt`.

File: src/ObjCRuntime/Dlfcn.cs [hunk 2]
```diff
@@ -240,6 +280,23 @@ public static nfloat GetNFloat (IntPtr handle, string symbol)
...
 		}
+
+		public static void SetNFloat (IntPtr handle, string symbol, nfloat value)
+		{
+			var indirect = dlsym (handle, symbol);
+			if (indirect == IntPtr.Zero)
+				return;
+
+			unsafe {
+				if (sizeof (IntPtr) == 4) {
```
Comment (rolfbjarne) [inline src/ObjCRuntime/Dlfcn.cs]: This doesn't work?

```csharp
nfloat *ptr = (nfloat *) indirect;
*ptr = value;
```

File: src/generator.cs [hunk 3]
```diff
@@ -6056,6 +6071,31 @@ public void Generate (Type type)
...
+						var btype = field_pi.PropertyType.GetEnumUnderlyingType ();
+						if (smartEnumTypeName != null) {
+							print ("if (_{0} == null)", field_pi.Name);
+							indent++;
+							print ("_{0} = Dlfcn.GetStringConstant (Libraries.{2}.Handle, \"{1}\");", field_pi.Name, fieldAttr.SymbolName, library_name);
+							indent--;
+							print ($"return {smartEnumTypeName}Extensions.GetValue (_{field_pi.Name});");
+						} else if (UnifiedAPI && IsNativeEnum (field_pi.PropertyType)) {
+							if (btype == TypeManager.System_nint || (BindThirdPartyLibrary && btype == TypeManager.System_Int64))
+								print ($"return ({fieldTypeName}) (int) Dlfcn.GetNInt (Libraries.{library_name}.Handle, \"{fieldAttr.SymbolName}\");" );
```
Comment (rolfbjarne) [inline src/generator.cs]: Should't this cast to long/ulong before casting to the enum?

```csharp
print ($"return ({fieldTypeName}) (long) Dlfcn.GetNInt (Libraries.{library_name}.Handle, \"{fieldAttr.SymbolName}\");" );
```

File: src/generator.cs [hunk 4]
```diff
@@ -6095,6 +6137,27 @@ public void Generate (Type type)
 							print ("Dlfcn.SetNFloat (Libraries.{2}.Handle, \"{1}\", value);", field_pi.Name, fieldAttr.SymbolName, library_name);
 						} else if (UnifiedAPI && field_pi.PropertyType == TypeManager.CoreGraphics_CGSize) {
 							print ("Dlfcn.SetCGSize (Libraries.{2}.Handle, \"{1}\", value);", field_pi.Name, fieldAttr.SymbolName, library_name);
+						} else if (field_pi.PropertyType.IsEnum) {
+							var btype = field_pi.PropertyType.GetEnumUnderlyingType ();
+							if (smartEnumTypeName != null)
+								print ($"Dlfcn.SetString (Libraries.{library_name}.Handle, \"{fieldAttr.SymbolName}\", value.GetConstant ());");
+							else if (UnifiedAPI && IsNativeEnum (field_pi.PropertyType)) {
+								if (btype == TypeManager.System_nint || (BindThirdPartyLibrary && btype == TypeManager.System_Int64))
+									print ($"Dlfcn.SetNInt (Libraries.{library_name}.Handle, \"{fieldAttr.SymbolName}\", (nint) (int) value);");
```
Comment (rolfbjarne) [inline src/generator.cs]: Same: this should cast to long/long before casting to nint/nuint:

```csharp
print ($"Dlfcn.SetNInt (Libraries.{library_name}.Handle, \"{fieldAttr.SymbolName}\", (nint) (long) value);");
```

File: src/generator.cs [hunk 5]
```diff
@@ -6056,6 +6071,31 @@ public void Generate (Type type)
...
+							else
+								throw new BindingException (1014, true, "Unsupported type for Fields: {0}", fieldTypeName);
+						} else {
+							if (btype == TypeManager.System_Int32)
+								print ($"return ({fieldTypeName}) Dlfcn.GetInt32 (Libraries.{library_name}.Handle, \"{fieldAttr.SymbolName}\");");
+							else if (btype == TypeManager.System_UInt32)
+								print ($"return ({fieldTypeName}) Dlfcn.GetUInt32 (Libraries.{library_name}.Handle, \"{fieldAttr.SymbolName}\");");
+							else if (btype == TypeManager.System_Int64)
+								print ($"return ({fieldTypeName}) Dlfcn.GetInt64 (Libraries.{library_name}.Handle, \"{fieldAttr.SymbolName}\");");
+							else
```
Comment (rolfbjarne) [inline src/generator.cs]: What about `ulong`?

File: src/ObjCRuntime/Dlfcn.cs (hunk 2 above)
Comment (dalexsoto) [inline src/ObjCRuntime/Dlfcn.cs]: yup tested locally that works, thanks!

File: src/generator.cs (hunk 3 above)
Comment (dalexsoto) [inline src/generator.cs]: Indeed my bad, this needs `long` instead of `int`

File: src/generator.cs (hunk 5 above)
Comment (dalexsoto) [inline src/generator.cs]: Coming soon to the theater next to you! ;)

File: src/ObjCRuntime/Dlfcn.cs [hunk 6]
```diff
@@ -222,11 +238,35 @@ public static nint GetNInt (IntPtr handle, string symbol)
...
+		}
+
 		public static nuint GetNUInt (IntPtr handle, string symbol)
 		{
 			return (nuint)GetIntPtr (handle, symbol);
 		}
 
+		public static void SetNUInt (IntPtr handle, string symbol, nuint value)
+		{
+			var indirect = dlsym (handle, symbol);
```
Comment (spouliot) [inline src/ObjCRuntime/Dlfcn.cs]: Same comment as Rolf above

**Comment by dalexsoto** [comment]
@rolfbjarne Fixed your feedback, thanks!

**Pull request merged**